tests/share/CommonFunctions.sh

```diff
@@ -43,56 +83,82 @@ function _Prepare {
     _log "INFO" "Purging directory '${LOG_DIR}'"
     rm -rf ${LOG_DIR}/*
-    echo "Contents of the ${LOG_DIR}  after cleanup"
+    echo "Contents of the ${LOG_DIR} after cleanup"
     ls -la ${LOG_DIR}
   else
     _log "INFO" "Directory '${LOG_DIR}' must exist."
   fi
 
-  sleep 5
-
   _log "INFO" "Preparing default Siodb instance"
-  dd if=/dev/urandom of=/etc/siodb/instances/siodb/master_key bs=16 count=1
-  cp -f ${SCRIPT_DIR}/../share/public_key /etc/siodb/instances/siodb/initial_access_key
+  dd if=/dev/urandom of=/etc/siodb/instances/${SIODB_INSTANCE}/master_key bs=16 count=1
+  cp -f ${SCRIPT_DIR}/../share/public_key /etc/siodb/instances/${SIODB_INSTANCE}/initial_access_key
 }
 
 function _ShowSiodbProcesses {
   echo "===== Siodb Processes ====="
-  ps -aux | grep siodb
+  if [[ $(pgrep -c siodb) -gt 0 ]]; then
+    pgrep -a siodb
+  fi
   echo "==========================="
 }
 
 function _StartSiodb {
   _log "INFO" "Starting default Siodb instance"
-  ${SIODB_BIN}/siodb --instance siodb --daemon
-  sleep ${startup_timeout}
+  ${SIODB_BIN}/siodb --instance ${SIODB_INSTANCE} --daemon
+  numberEntriesInLog=0
+  counterTimeout=0
+  while [[ $numberEntriesInLog -eq 0 ]]; do
+    _log "INFO" "Waiting for instance to be ready..."
+    numberEntriesInLog=$(egrep 'Listening for (TCP|UNIX) connections' ${LOG_DIR}/siodb_*.log | wc -l | bc)
+    if [[ ${counterTimeout} -gt ${instanceStartupTimeout} ]]; then
+    _log "ERROR" "Timeout (${instanceStartupTimeout} seconds) reached while starting the instance..."
+    fi
+    counterTimeout=$((counterTimeout+1))
+    sleep 1
+  done
   _ShowSiodbProcesses
-  sleep 3
 }
 
 function _StopSiodb {
-  _log "INFO" "Stopping Siodb process on default instance"
-  _ShowSiodbProcesses
-  SIODB_PROCESS_ID=$(ps -ef | grep 'siodb --instance siodb --daemon' | grep -v grep \
-    | awk '{print $2}')
-  if [[ -z "${SIODB_PROCESS_ID}" ]]; then
-    echo "Siodb is already not running."
+  if [[ "${SIOTEST_KEEP_INSTANCE_UP}" == "0" ]]; then
+    _log "INFO" "Stopping Siodb process on default instance"
+    _ShowSiodbProcesses
+    SIODB_PROCESS_ID=$(ps -ef | grep "siodb --instance ${SIODB_INSTANCE} --daemon" | grep -v grep \
+      | awk '{print $2}')
+    if [[ -z "${SIODB_PROCESS_ID}" ]]; then
+      echo "Siodb is already not running."
+    else
+      kill -SIGINT ${SIODB_PROCESS_ID}
+      counterSiodbProcesses=$(ps -ef | grep "siodb --instance ${SIODB_INSTANCE} --daemon" | grep -v grep \
+        | awk '{print $2}' | wc -l | bc)
+      counterTimeout=0
+      while [[ $counterSiodbProcesses -ne 0 ]]; do
+        _log "INFO" "Waiting for instance to be stopped..."
+        counterSiodbProcesses=$(ps -ef | grep "siodb --instance ${SIODB_INSTANCE} --daemon" | grep -v grep \
+        | awk '{print $2}' | wc -l | bc)
+        if [[ ${counterTimeout} -gt ${instanceStartupTimeout} ]]; then
+          _log "ERROR" "Timeout (${instanceStartupTimeout} seconds) reached while stopping the instance..."
+        fi
```

Comment:
I think at this point you need to `kill -9` instance and break out of the loop

Reply:
With _log,  if _log == "ERROR" then it exits 1.

Reply:
This is absolutely not obvious that script does this when logging. All code must be readable and understandable, even scripting code. Please remove such functionality and do explicit `exit` after logging error. Alternative could be renaming `_log` into something like `logAndMaybeExit`, but this is too long and complicated name, does not add readability.
So once again, make `_log` just log something and that's all it should do. Use explicit `exit` to exit on error.